### PR TITLE
Reduction of unnecessary downloads from GPU

### PR DESCRIPTION
**Pull request by alphaville**
## Main Changes

- Extra checks only if `GPUTILS_DEBUG_MODE` is defined
- methods do not return a status code now
- status code accessible with a getter (docs written)

## Associated Issues

None

### CODE REVIEW

File: include/tensor.cuh
Comment (ruairimoran) [inline include/tensor.cuh:1]: After line 2035 (in updated code), we need `TEMPLATE_CONSTRAINT_REQUIRES_FPX` to remove `Requires clause differs in template redeclaration` warning

File: include/tensor.cuh [hunk 1]
```diff
@@ -1225,8 +1254,14 @@ private:
      */
     void computeWorkspaceSize(size_t m, size_t n);
 
+
 public:
 
+    int getStatus() {
```
Comment (ruairimoran) [inline include/tensor.cuh]: should this not be `statusCode()`

File: include/tensor.cuh [hunk 2]
```diff
@@ -1225,8 +1254,14 @@ private:
...
+        // redefinition of getStatus
+        return m_svdStatus;
```
Comment (ruairimoran) [inline include/tensor.cuh]: why not have `m_svdStatus = std::max(m_svdStatus, (*m_info)(0, 0, 0));` here, so you don't need the `#ifdef GPUTILS_DEBUG_MODE` blocks ? 

Why even have m_svdStatus ? We can redefine `statusCode()` with `return (*m_info)(0, 0, 0)`. You will want to call `statusCode()` after every method when debugging anyway, so you catch the first instance of a problem

**Review by ruairimoran — COMMENTED** [review]
some minor comments. Also, there appears to be a memory leak

File: include/tensor.cuh (hunk 2 above)
Comment (alphaville) [reply include/tensor.cuh]: I've changed the code slightly:

1. In `CholeskyBatchFactoriser` we need to allocate more memory for the device info; this is required by cuBLAS; see API docs of  [`potrsbatched`](https://docs.nvidia.com/cuda/cusolver/index.html#cusolverdn-t-potrsbatched) 
2. In `Svd` we perform multiple calls of `cusolverDnDgesvd` in a for loop and we result `m_info`. If the maximum of all these `m_info` is zero, all is good. 

Anyway, let's talk about this on Teams.

**Review by ruairimoran — APPROVED** [review]
lgtm

**Pull request merged**